File: output_20260318_091315/claude-opus/java/java/Linklist/Spouse.java
class Person {

  protected Person spouse;

  /*@
  predicate person_spouse(Person p; Person s) =
    this.spouse |-> s &*&
    (s == null || s.spouse |-> this);
  @*/

  //@ requires person_spouse(this, spouse);
  //@ ensures person_spouse(this, spouse);
  public void spouse_symm()
  {
    //@ open person_spouse(this, spouse);
    //@ open person_spouse(spouse, spouse.spouse);
    //@ close person_spouse(this, spouse);
    //@ close person_spouse(spouse, spouse.spouse);
  }

  public Person()
  //@ ensures person_spouse(this, null);
  {
    spouse = null;
    //@ close person_spouse(this, null);
  }

  //@ requires person_spouse(this, spouse);
  //@ ensures person_spouse(this, spouse) &*& result == spouse;
  public Person getSpouse()
  {
    //@ open person_spouse(this, spouse);
    Person result = spouse;
    //@ close person_spouse(this, spouse);
    return result;
  }

  //@ requires person_spouse(this, spouse) &*& person_spouse(other, other.spouse);
  //@ ensures person_spouse(this, other) &*& person_spouse(other, this);
  protected void setSpouse(Person other)
  {
    //@ open person_spouse(this, spouse);
    //@ open person_spouse(other, other.spouse);
    spouse = other;
    other.spouse = this;
    //@ close person_spouse(this, other);
    //@ close person_spouse(other, this);
  }

  //@ requires person_spouse(this, spouse) &*& spouse != null &*& person_spouse(spouse, this);
  //@ ensures person_spouse(this, null) &*& person_spouse(spouse, null);
  protected void clearSpouse()
  {
    //@ open person_spouse(this, spouse);
    //@ open person_spouse(spouse, this);
    spouse.spouse = null;
    spouse = null;
    //@ close person_spouse(spouse, null);
    //@ close person_spouse(this, null);
  }

  //@ requires person_spouse(this, spouse) &*& person_spouse(other, other.spouse);
  //@ ensures person_spouse(this, other) &*& person_spouse(other, this);
  void marry(Person other)
  {
    other.setSpouse(this);
  }

  //@ requires person_spouse(this, spouse) &*& spouse != null &*& person_spouse(spouse, this);
  //@ ensures person_spouse(this, null) &*& person_spouse(spouse, null);
  void divorce()
  {
    spouse.clearSpouse();
  }

}

class Program {

  public static void main(String[] args)
  //@ requires true;
  //@ ensures true;
  {
    Person a = new Person();
    Person b = new Person();
    a.marry(b);
    b.divorce();
  }

}

File: output_20260321_024012/gpt52/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@
predicate person(struct person *p; struct person *s) =
  p != 0 &*& malloc_block_person(p) &*& p->spouse |-> s;

predicate person0(struct person *p) = person(p, _);

predicate persons2(struct person *a, struct person *b) =
  person(a, b) &*& person(b, a);

predicate persons2_divorced(struct person *a, struct person *b) =
  person(a, 0) &*& person(b, 0);

predicate persons3_all(struct person *a, struct person *b, struct person *c) =
  person0(a) &*& person0(b) &*& person0(c);
@*/

struct person *create_person()
  //@ requires true;
  //@ ensures person(result, 0);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  //@ assume(p != 0);
  p->spouse = 0;
  //@ close person(p, 0);
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires person(this, 0) &*& person(other, 0);
  //@ ensures persons2(this, other);
{
  //@ open person(this, 0);
  //@ open person(other, 0);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this, other);
  //@ close person(other, this);
}

struct person* person_get_spouse(struct person* this)
  //@ requires person0(this);
  //@ ensures person0(this) &*& result == 0 ? true : true;
{
  //@ open person(this, _);
  struct person* s = this->spouse;
  //@ close person(this, s);
  return s;
}

void divorce(struct person* this)
  //@ requires exists<struct person*>(?other) &*& persons2(this, other);
  //@ ensures persons2_divorced(this, other);
{
  //@ open exists(other);
  //@ open person(this, other);
  //@ open person(other, this);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close person(this, 0);
  //@ close person(other, 0);
}

void die(struct person *this)
  //@ requires person(this, ?s) &*& (s == 0 ? true : person(s, this));
  //@ ensures s == 0 ? true : person(s, 0);
{
  //@ open person(this, s);
  if(this->spouse != 0) {
    //@ open person(s, this);
    this->spouse->spouse = 0;
    //@ close person(s, 0);
  }
  free(this); 
}

int main() 
  //@ requires true;
  //@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  //@ close exists(bob);
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  //@ close exists(eve);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}